### PR TITLE
ODQA Reader

### PR DESCRIPTION
**Pull request by my-master**
Key features:

* added `ODQADataReader` class
    * parallel SQLite DB building from usual text data
    * get rid of DrQa reference for DB building
    * updated docs

### CODE REVIEW

**Comment by my-master** [comment]
Made changes.

File: deeppavlov/dataset_iterators/sqlite_iterator.py [hunk 1]
```diff
@@ -53,15 +51,34 @@ class SQLiteDataIterator(DataFittingIterator):
...
-        download_dir = expand_path(data_dir)
-        download_path = download_dir.joinpath(data_url.split("/")[-1])
-        download(download_path, data_url, force_download=False)
-
-        self.connect = sqlite3.connect(str(download_path), check_same_thread=False)
-        self.db_name = self.get_db_name()
+        if load_path is not None:
+            if load_path.startswith('http'):
+                logger.info("Downloading database from url: {}".format(load_path))
+                download_dir = expand_path(Path(load_path)).parent
```
Comment (yoptar) [inline deeppavlov/dataset_iterators/sqlite_iterator.py]:
```python
expand_path(Path('http://files.deeppavlov.ai/datasets/wikipedia/enwiki.db')).parent
```
returns `PosixPath('/home/yoptar/reps/DeepPavlov/download/http:/files.deeppavlov.ai/datasets/wikipedia')` which is not ideal

File: deeppavlov/dataset_readers/odqa_reader.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        logger.info('Reading files...')
+        save_path = expand_path(kwargs['save_path'])
+        if save_path.exists() and Path(save_path.parent / '.done').exists():
+            return
+
+        if db_url:
+            if Path(save_path).is_file():
+                download_dir = save_path.parent
+            else:
+                download_dir = save_path
```
Comment (yoptar) [inline deeppavlov/dataset_readers/odqa_reader.py]: `Path(save_path).is_file()` will return `True` only if a file already exists. So it should not determine download directory

File: deeppavlov/dataset_readers/odqa_reader.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+        Raises:
+            sqlite3.OperationalError if `save_path` doesn't exist.
+            RuntimeError if dataset_format is not in ['txt', 'json', 'wiki']
+
+        Returns:
+            None
+
+        """
+        done_path = Path(save_path.parent / '.done')
```
Comment (yoptar) [inline deeppavlov/dataset_readers/odqa_reader.py]: `save_path.with_suffix(save_path.suffix + '.done')`
will create a unique `.done` file for every file name

File: deeppavlov/dataset_readers/odqa_reader.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+        if Path(save_path).exists():
+            Path(save_path).unlink()
+        if done_path.exists():
+            done_path.unlink()
+
+        logger.info('Building the database...')
+
+        try:
+            conn = sqlite3.connect(str(save_path))
```
Comment (yoptar) [inline deeppavlov/dataset_readers/odqa_reader.py]: I think we should not punish users for not wanting to create directories by hand.
`save_path.parent.mkdir(parents=True, exist_ok=True)`

File: deeppavlov/dataset_readers/odqa_reader.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        """Extract file contents from '.txt' file.
+
+        Args:
+            fpath: path to a '.txt' file.
+
+        Returns:
+             a list with tuple of normalized file name and file contents
+
+        """
+        with open(fpath) as fin:
```
Comment (yoptar) [inline deeppavlov/dataset_readers/odqa_reader.py]: encoding needs to be set explicitly as default value differs between different OS's

File: deeppavlov/dataset_readers/odqa_reader.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+        Args:
+            fpath: path to a '.json' file.
+
+        Returns:
+            a list with tuples of normalized file name and file contents
+
+        """
+        docs = []
+        with open(fpath) as fin:
```
Comment (yoptar) [inline deeppavlov/dataset_readers/odqa_reader.py]: encoding

File: deeppavlov/dataset_readers/odqa_reader.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+        Args:
+            fpath: path to a '.txt' file in wiki extractor format
+
+        Returns:
+            a list with tuples of normalized file name and file contents
+
+        """
+        docs = []
+        with open(fpath) as fin:
```
Comment (yoptar) [inline deeppavlov/dataset_readers/odqa_reader.py]: encoding

File: docs/components/tfidf_ranking.rst [hunk 8]
```diff
@@ -125,15 +137,14 @@ and is built by the following steps:
...
-   script <https://github.com/facebookresearch/DrQA/blob/master/scripts/retriever/build_db.py>`__.
+#. Build a database during :ref:`ranker_training`.
 
 enwiki_tfidf_matrix.npz
 -------------------------
 
 **enwiki_tfidf_matrix.npz** is a full Wikipedia tf-idf matrix of
 size **hash_size x number of documents** which is
-**2**24 x 5159530**. This matrix is built with
+**2**24 x 5180368**. This matrix is built with
```
Comment (yoptar) [inline docs/components/tfidf_ranking.rst]: Maybe replace __2**24__ with __2²⁴__? I do not insist but it looks better in the doc in my opinion
image

File: docs/components/tfidf_ranking.rst [hunk 9]
```diff
@@ -17,42 +17,52 @@ Config
...
 Default ranker config for **English** language is
-:config:`doc_retrieval/en_ranker_tfidf_wiki.json <ranking/en_ranker_tfidf_wiki.json>`
+:config:`doc_retrieval/en_ranker_tfidf_wiki.json <doc_retrieval/en_ranker_tfidf_wiki.json>`
 
 Default ranker config for **Russian** language is
-:config:`doc_retrieval/ru_ranker_tfidf_wiki.json <ranking/ru_ranker_tfidf_wiki.json>`
+:config:`doc_retrieval/ru_ranker_tfidf_wiki.json <doc_retrieval/ru_ranker_tfidf_wiki.json>`
 
 Config Structure
 ----------------
```
Comment (nikolay-bushkov) [inline docs/components/tfidf_ranking.rst]: We already have "Configuration files" doc page describing the general config structure. Consider writing your doc page assuming the user knows that; just add some reference to the "Configuration files" and corresponding API doc.

File: deeppavlov/dataset_readers/odqa_reader.py (hunk 2 above)
Comment (my-master) [reply deeppavlov/dataset_readers/odqa_reader.py]: ok

File: deeppavlov/dataset_readers/odqa_reader.py (hunk 3 above)
Comment (my-master) [reply deeppavlov/dataset_readers/odqa_reader.py]: cool

File: deeppavlov/dataset_readers/odqa_reader.py (hunk 4 above)
Comment (my-master) [reply deeppavlov/dataset_readers/odqa_reader.py]: I think we should not punish users for making typos either. However, will do as accepted by voting.

File: deeppavlov/dataset_readers/odqa_reader.py (hunk 5 above)
Comment (my-master) [reply deeppavlov/dataset_readers/odqa_reader.py]: ok

File: deeppavlov/dataset_readers/odqa_reader.py (hunk 6 above)
Comment (my-master) [reply deeppavlov/dataset_readers/odqa_reader.py]: ok

File: deeppavlov/dataset_readers/odqa_reader.py (hunk 7 above)
Comment (my-master) [reply deeppavlov/dataset_readers/odqa_reader.py]: ok

File: deeppavlov/dataset_iterators/sqlite_iterator.py [hunk 10]
```diff
@@ -53,15 +50,23 @@ class SQLiteDataIterator(DataFittingIterator):
 
     """
 
-    def __init__(self, data_dir: str = '', data_url: str = DB_URL, batch_size: int = None,
+    def __init__(self, load_path: Union[str, Path], batch_size: int = None,
```
[the remark this answers is not in the document]
Comment (my-master) [reply deeppavlov/dataset_iterators/sqlite_iterator.py]: sure

File: deeppavlov/dataset_iterators/sqlite_iterator.py [hunk 11]
```diff
@@ -53,15 +50,23 @@ class SQLiteDataIterator(DataFittingIterator):
...
                  shuffle: bool = None, seed: int = None, **kwargs):
 
-        download_dir = expand_path(data_dir)
-        download_path = download_dir.joinpath(data_url.split("/")[-1])
-        download(download_path, data_url, force_download=False)
-
-        self.connect = sqlite3.connect(str(download_path), check_same_thread=False)
-        self.db_name = self.get_db_name()
```
[the remark this answers is not in the document]
Comment (my-master) [reply deeppavlov/dataset_iterators/sqlite_iterator.py]: Why?

File: deeppavlov/vocabs/wiki_sqlite.py [hunk 12]
```diff
@@ -36,8 +36,8 @@ class WikiSQLiteVocab(SQLiteDataIterator):
 
     """
 
-    def __init__(self, data_url: str, data_dir: str = '', join_docs: bool=True, shuffle: bool=False, **kwargs):
-        super().__init__(data_dir=data_dir, data_url=data_url, shuffle=shuffle)
+    def __init__(self, load_path: str, join_docs: bool=True, shuffle: bool=False, **kwargs):
```
[the remark this answers is not in the document]
Comment (my-master) [reply deeppavlov/vocabs/wiki_sqlite.py]: Do we explicitly point the return value in `__init__()` methods? Most of the library's classes don't do so.

File: deeppavlov/dataset_iterators/sqlite_iterator.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (nikolay-bushkov) [reply deeppavlov/dataset_iterators/sqlite_iterator.py]: Nevermind...

File: deeppavlov/vocabs/wiki_sqlite.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (nikolay-bushkov) [reply deeppavlov/vocabs/wiki_sqlite.py]: Is `__init__()` special enough to break the rules?
I think now the state of this kind of annotation is about 50/50... and we are continuously improving

File: docs/components/tfidf_ranking.rst (hunk 9 above)
Comment (my-master) [reply docs/components/tfidf_ranking.rst]: ok

**Pull request merged**